### PR TITLE
Package: Webpack plugin to externalize WordPress deps

### LINKED ISSUE #11907: Automatically export and use list of `@wordpress` dependencies on build
We currently have a disconnect between knowing what `@wordpress` dependencies Gutenberg extensions actually depend on and what we declare as dependencies when enqueueing these scripts at Jetpack.

This is error prone as demonstrated in https://github.com/Automattic/wp-calypso/pull/28752 that I had to revert.

This will get more fragile as we'll have more blocks to test and `@wordpress` dependencies will sneak in via imports from Calypso.

We might also be enqueuing wp-scripts that we don't actually depend on anymore.

E2E tests will help in future to catch worst issues but developer experience of keeping Jetpack and blocks in sync isn't great. It complicates testing as testers need to run blocks from Calypso branch against specific Jetpack branches. 

I'd suggest we:
- Collect `@wordpress` dependencies (+`lodash`) during the build.
- Export them via JSON object such as the manifest file.
- Read dependencies from that JSON file in Jetpack.

### LINKED ISSUE #31486: Blocks: Spike on webpack plugin to externalize WP deps
- Externalize known WP deps
- Provide a json describing the WP deps

Something like:

```js
// Webpack plugin config
new ExternalizeWpScriptDeps( someConfig );

// entry.js
import { Component } from '@wordpress/element';

// output.js
// …whatever that webpack magic looks like
const Component = wp.element;

// output.deps.json
[ 'wp-element' ]
```

Enqueue would be aware of WP deps, no more disconnect!

```php
wp_enqueue_script( 'entry.js', json_decode( 'path/to/output.deps.json' ) );
```

### PR DESCRIPTION
**Pull request by sirreal**
Introduce a webpack plugin as described in #31486
Partly resolves https://github.com/Automattic/jetpack/issues/11907

https://github.com/Automattic/jetpack/pull/11591 demonstrates how the output of this plugin would be used:

```php
$script_relative_path = 'path/to/script.js';
$script_deps_path     = 'path/to/script.deps.json';
$script_dependencies = file_exists( $script_deps_path )
	? json_decode( file_get_contents( $script_deps_path ) )
	: array();
$view_script = plugins_url( $script_relative_path, __FILE__ );
wp_enqueue_script( 'script', $view_script, $script_dependencies );
```

---

Example usage:

```js
// Webpack config
{
  // …
  plugins: [
    new WordPressExternalDependenciesPlugin(),
  ]
}

// entry.js
import { Component } from '@wordpress/element';

// output.js
// …whatever that webpack magic looks like
const Component = wp.element;

// output.deps.json
[ 'wp-element' ]
```

Closes #31486

## To do
- ❌ Shall we always include `wp-polyfill` or make that an option? (handle in follow-up)

## Testing

I've put the plugin into use for the Jetpack blocks build. The entry points like `editor.js` should produce `editor.deps.json`.

- `npx lerna run prepare --scope="*/o2-blocks"` - is an `editor.deps.json` built to the same location as `editor.js`?
- Test in Gutenberg following [these instructions](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946). They can be adapted for use in Jetpack as well.
- Is a `.deps.json` file created for each entry point?
- Do the deps look correct? You can compare them with our manually managed lists:
  - [Editor](https://github.com/Automattic/jetpack/blob/3b38a64e0eca52e618245230ca2eb13e39191094/class.jetpack-gutenberg.php#L516-L542)
  - [Map](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/map/map.php#L25-L34)
  - [Tiled gallery](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/tiled-gallery/tiled-gallery.php#L29-L33)
  - [Slidshow](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/slideshow/slideshow.php#L25-L30)
  - …
- The blocks should continue to work exactly as before. `@wordpress`, `jquery`, `react`, `react-dom`, `moment`, and `lodash` deps should be externals that appear in _exactly the same way_ as before.

### CODE REVIEW

**Comment by simison** [comment]
What do you think about including other non-wp externals shipped in WordPress here?

List used by Gutenberg:

https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L46-L64

File: packages/wordpress-external-dependencies-plugin/index.js
Comment (simison) [inline packages/wordpress-external-dependencies-plugin/index.js]: I like `@wordpress/api-fetch` example more here because it also shows also how we camelCase in this function.

Example doc block: https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L13-L28

**Comment by simison** [comment]
>What do you think about including other non-wp.* externals shipped in WordPress here?

Related naive question; should we then need to worry about aliases such as `lodash-es` → `lodash`?

https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L85-L89

**Comment by sirreal** [comment]
> > What do you think about including other non-wp.* externals shipped in WordPress here?
> 
> Related naive question; should we then need to worry about aliases such as `lodash-es` → `lodash`?
> 
> https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L85-L89

We should reasonably be able to cover a lot of cases and it's good to keep them in mind. [It concerns me how varied lodash imports might look](https://github.com/lodash/lodash#installation_):

- `lodash`
- `lodash/map`
- `lodash-es`
- `lodash-es/map`?
- [`lodash.map`](https://www.npmjs.com/package/lodash.map)
- `lodash/fp` (likely we won't support this)
- `lodash/fp/map` (likely we won't support this)

I'd like to focus on supporting the basics on a first pass. In the future,

File: packages/wordpress-external-dependencies-plugin/index.js
Comment (jsnajdr) [inline packages/wordpress-external-dependencies-plugin/index.js]: `<nitpick>`
The modern JS style is to prefer composition over inheritance:
```js
class MyPlugin {
  constructor() {
    this.innerPlugin = new ExternalsPlugin( 'global', [ wpExternals ] );
  }

  apply( compiler ) {
    this.innerPlugin.apply( compiler );
  }
}
```
One benefit is that it's clear what the plugin's public API is.
`</nitpick>`

File: packages/wordpress-external-dependencies-plugin/package.json
Comment (jsnajdr) [inline packages/wordpress-external-dependencies-plugin/package.json]: This is a webpack plugin that's ever run only from Node.js, no browser or bundler is ever involved. The`sideEffects` flag is not needed and `main` has default value anyway.

**Comment by sirreal** [comment]
The `jetpack-blocks` output isn't the same and I believe it's because additional things are externed like `react` and `moment.

I've been running the following to produce non-minified output to diff:

```sh
git checkout master
npm ci && npx lerna bootstrap --concurrency=2 --scope '@automattic/jetpack-blocks' && npx lerna run build --stream --scope '@automattic/jetpack-blocks'
mv packages/jetpack-blocks/dist packages/jetpack-blocks/master-dist

git checkout add/package-wp-extern-plugin
npm ci && npx lerna bootstrap --concurrency=2 --scope '@automattic/jetpack-blocks' && npx lerna run build --stream --scope '@automattic/jetpack-blocks'

# diff the output

kdiff3 packages/jetpack-blocks/master-dist packages/jetpack-blocks/dist
```

**Review by ockham — DISMISSED** [review]
Thanks Jon, awesome job! :tada: 

> Shall we always include `wp-polyfill` or make that an option?

Not sure. Might be fine to always include it for a first iteration and potentially add the option later if it turns out that we need it. (We'd need to add the polyfill tho before actually using the `deps` files in JP, i.e. before merging https://github.com/Automattic/jetpack/pull/11591.)

> The `jetpack-blocks` output isn't the same and I believe it's because additional things are externed like `react` and `moment`.

I tend to agree, from a cursory glance at the diff. The blocks I've tested seem to work well -- if there's anything off, I'm confident that we'll discover it before the next release.

File: packages/wordpress-external-dependencies-plugin/package.json
Comment (simison) [reply packages/wordpress-external-dependencies-plugin/package.json]: Should docs then be updated? It says:

> Your package.json can have any of the normal properties but at a minimum should contain `main`, `module`, and `sideEffects`.

https://github.com/Automattic/wp-calypso/blob/2b2ca2b01466b33add52c533cb6383cef7080be2/docs/monorepo.md

**Comment by sirreal** [comment]
I tried running this plugin in the Gutenberg webpack config and it exposed some issues. I've added some hacks that make the synthetic `.deps.json` asset look like a chunk or module so that an appropriate path can be built.

I'd be open to more idiomatic ways of creating a chunk. Source map plugins should provide some insight, but my comprehension of webpack internals is still quite limited.

To try it agains [Gutenberg](https://github.com/WordPress/gutenberg), you'll need to link this package (`npm link` from the package dir and `npm link @automattic/wordpress-external-dependencies-plugin` from the Gutenberg root) and apply the patch below to Gutenberg. Then `npm install` and `npm run build`.

<details>
<summary>Gutenberg patch</summary>

```patch
diff --git a/packages/scripts/config/webpack.config.js b/packages/scripts/config/webpack.config.js
index 9a1daabf5..516f6a907 100644
--- a/packages/scripts/config/webpack.config.js
+++ b/packages/scripts/config/webpack.config.js
@@ -4,54 +4,16 @@
 const { BundleAnalyzerPlugin } = require( 'webpack-bundle-analyzer' );
 const LiveReloadPlugin = require( 'webpack-livereload-plugin' );
 const path = require( 'path' );
+const WordPressExternalDependenciesPlugin = require("@automattic/wordpress-external-dependencies-plugin");
 
 /**
  * Internal dependencies
  */
 const { camelCaseDash, hasBabelConfig } = require( '../utils' );
 
-/**
- * Converts @wordpress/* string request into request object.
- *
- * Note this isn't the same as camel case because of the
- * way that numbers don't trigger the capitalized next letter.
- *
- * @example
- * formatRequest( '@wordpress/api-fetch' );
- * // { this: [ 'wp', 'apiFetch' ] }
- * formatRequest( '@wordpress/i18n' );
- * // { this: [ 'wp', 'i18n' ] }
- *
- * @param {string} request Request name from import statement.
- * @return {Object} Request object formatted for further processing.
- */
-const formatRequest = ( request ) => {
-	// '@wordpress/api-fetch' -> [ '@wordpress', 'api-fetch' ]
-	const [ , name ] = request.split( '/' );
-
-	// { this: [ 'wp', 'apiFetch' ] }
-	return {
-		this: [ 'wp', camelCaseDash( name ) ],
-	};
-};
-
-const wordpressExternals = ( context, request, callback ) => {
-	if ( /^@wordpress\//.test( request ) ) {
-		callback( null, formatRequest( request ), 'this' );
-	} else {
-		callback();
-	}
-};
 
 const externals = [
 	{
-		react: 'React',
-		'react-dom': 'ReactDOM',
-		moment: 'moment',
-		jquery: 'jQuery',
-		lodash: 'lodash',
-		'lodash-es': 'lodash',
-
 		// Distributed NPM packages may depend on Babel's runtime regenerator.
 		// In a WordPress context, the regenerator is assigned to the global
 		// scope via the `wp-polyfill` script. It is reassigned here as an
@@ -60,7 +22,6 @@ const externals = [
 		// See: https://github.com/WordPress/gutenberg/issues/13890
 		'@babel/runtime/regenerator': 'regeneratorRuntime',
 	},
-	wordpressExternals,
 ];
 
 const isProduction = process.env.NODE_ENV === 'production';
@@ -111,6 +72,7 @@ const config = {
 		// WP_LIVE_RELOAD_PORT global variable changes port on which live reload works
 		// when running watch mode.
 		! isProduction && new LiveReloadPlugin( { port: process.env.WP_LIVE_RELOAD_PORT || 35729 } ),
+		new WordPressExternalDependenciesPlugin(),
 	].filter( Boolean ),
 	stats: {
 		children: false,
```

</details>

File: packages/wordpress-external-dependencies-plugin/README.md
Comment (sirreal) [inline packages/wordpress-external-dependencies-plugin/README.md]: Need a better README

File: packages/wordpress-external-dependencies-plugin/wp-modules.js
Comment (simison) [reply packages/wordpress-external-dependencies-plugin/wp-modules.js]: Why not generate `@wordpress/` deps instead of defining each one manually, since their naming patterns are predictable and there seems to be a kind of contract that they'll remain to be so?

Comment (sirreal) [reply packages/wordpress-external-dependencies-plugin/wp-modules.js]: My thinking was that there are `@wordpress/` packages that aren't available as script dependencies in WordPress and would need to be bundled. It may be that those packages are not for the browser and this bundler shouldn't need to worry about them.

Comment (sirreal) [reply packages/wordpress-external-dependencies-plugin/wp-modules.js]: This would be an interesting configuration point, allow consumers to specify the elements of this map so if they're designing a plugin that exposes additional WordPress script dependencies, they can define them here.

Comment (gziolo) [reply packages/wordpress-external-dependencies-plugin/wp-modules.js]: It very impractical because there will be more packages created and published and you will have to remember to update it.

**Comment by gziolo** [comment]
I don't quite understand why you need to manually craft this list of dependencies. You probably could auto-generate it based on npm dependencies added to your `package.json` file or with a similar trick. This one of the examples of how we automate the maintenance of packages in Gutenberg:
https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L26-L28

**Comment by jsnajdr** [comment]
The Gutenberg trick in particular will probably stop working as soon as some `@wordpress/*` dependency is transitive. For example, our `package.json` can specify a `@wordpress/data` dependency, but the fact that it further depends on `@wordpress/api-fetch` is hidden.

We need a better trick 🙂

**Comment by sirreal** [comment]
> The Gutenberg trick in particular will probably stop working as soon as some `@wordpress/*` dependency is transitive. For example, our `package.json` can specify a `@wordpress/data` dependency, but the fact that it further depends on `@wordpress/api-fetch` is hidden.
> 
> We need a better trick 🙂

We were burned by a transitive dependency on wp-polyfill recently, but I think that's a special case.

This would be targeting the WordPress environment where these scripts already declare their dependencies, so I don't think transitive dependencies are a concern for this case, unless I'm missing something.

> I don't quite understand why you need to manually craft this list of dependencies.

We don't necessarily need to do that, but was a way of getting started 🙂 

`@wordpress/*` dependencies may have consistent ways to convert to and from npm, global name, and the script name in WordPress, but not all the dependencies do.

**Comment by roccotripaldi** [comment]
Rebased, and re-ran the diffs with the new webpack.config in jetpack blocks.
There were a lot of differences, but i believe that's expected?

Give it an other spin before merging.

**Comment by sirreal** [comment]
> Give it an other spin before merging.

Hold off for now please, I want to iterate a bit based on initial feedback before landing this 🙂

**Comment by sirreal** [comment]
I've build Gutenberg with this plugin [using the instructions above](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946), full results below are very interesting to compare with [the declared script deps in core](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L268-L494).

One example:

`build/format-library/index.deps.json`:
```json
[
  "lodash",
  "wp-block-editor",
  "wp-components",
  "wp-element",
  "wp-i18n",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
```

[declared deps](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L439-L449):
```php
		'format-library'                     => array(
			'wp-block-editor',
			'wp-components',
			'wp-editor',
			'wp-element',
			'wp-i18n',
			'wp-keycodes',
			'wp-polyfill',
			'wp-rich-text',
			'wp-url',
		),
```

Notably in this example, [`lodash` is missing](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/js/dist/format-library.js#L764-L765):

```js
// EXTERNAL MODULE: external "lodash"
var external_lodash_ = __webpack_require__(2);
```

<details>

<summary>Full `deps.json` results</summary>

```
a11y
[
  "wp-dom-ready"
]
annotations
[
  "lodash",
  "wp-data",
  "wp-hooks",
  "wp-i18n",
  "wp-rich-text"
]
api-fetch
[
  "wp-i18n",
  "wp-url"
]
autop
[]
blob
[]
block-editor
[
  "lodash",
  "react",
  "wp-a11y",
  "wp-blob",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-deprecated",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-token-list",
  "wp-url",
  "wp-viewport",
  "wp-wordcount"
]
block-library
[
  "lodash",
  "moment",
  "wp-a11y",
  "wp-api-fetch",
  "wp-autop",
  "wp-blob",
  "wp-block-editor",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-date",
  "wp-deprecated",
  "wp-editor",
  "wp-element",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url",
  "wp-viewport"
]
block-serialization-default-parser
[]
block-serialization-spec-parser
[]
blocks
[
  "lodash",
  "wp-autop",
  "wp-blob",
  "wp-block-serialization-default-parser",
  "wp-compose",
  "wp-data",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-shortcode"
]
components
[
  "lodash",
  "moment",
  "react",
  "react-dom",
  "wp-a11y",
  "wp-api-fetch",
  "wp-compose",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
compose
[
  "lodash",
  "wp-element",
  "wp-is-shallow-equal"
]
core-data
[
  "lodash",
  "wp-api-fetch",
  "wp-data",
  "wp-deprecated",
  "wp-url"
]
data
[
  "lodash",
  "wp-compose",
  "wp-deprecated",
  "wp-element",
  "wp-is-shallow-equal",
  "wp-priority-queue",
  "wp-redux-routine"
]
date
[
  "moment"
]
deprecated
[
  "wp-hooks"
]
dom-ready
[]
dom
[
  "lodash"
]
edit-post
[
  "lodash",
  "wp-a11y",
  "wp-api-fetch",
  "wp-block-editor",
  "wp-block-library",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-editor",
  "wp-element",
  "wp-hooks",
  "wp-i18n",
  "wp-keycodes",
  "wp-notices",
  "wp-nux",
  "wp-plugins",
  "wp-url",
  "wp-viewport"
]
edit-widgets
[
  "wp-element"
]
editor
[
  "lodash",
  "react",
  "wp-api-fetch",
  "wp-autop",
  "wp-blob",
  "wp-block-editor",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-date",
  "wp-deprecated",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-keycodes",
  "wp-notices",
  "wp-nux",
  "wp-rich-text",
  "wp-url",
  "wp-viewport",
  "wp-wordcount"
]
element
[
  "lodash",
  "react",
  "react-dom",
  "wp-escape-html"
]
escape-html
[]
format-library
[
  "lodash",
  "wp-block-editor",
  "wp-components",
  "wp-element",
  "wp-i18n",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
hooks
[]
html-entities
[]
i18n
[]
is-shallow-equal
[]
keycodes
[
  "lodash",
  "wp-i18n"
]
list-reusable-blocks
[
  "lodash",
  "wp-api-fetch",
  "wp-components",
  "wp-compose",
  "wp-element",
  "wp-i18n"
]
notices
[
  "lodash",
  "wp-a11y",
  "wp-data"
]
nux
[
  "lodash",
  "wp-components",
  "wp-compose",
  "wp-data",
  "wp-element",
  "wp-i18n"
]
plugins
[
  "lodash",
  "wp-compose",
  "wp-element",
  "wp-hooks"
]
priority-queue
[]
redux-routine
[
  "lodash"
]
rich-text
[
  "lodash",
  "wp-compose",
  "wp-data",
  "wp-element",
  "wp-escape-html",
  "wp-hooks"
]
shortcode
[
  "lodash"
]
token-list
[
  "lodash"
]
url
[]
viewport
[
  "lodash",
  "wp-compose",
  "wp-data"
]
wordcount
[
  "lodash"
]
```

</details>

File: webpack.config.js
Comment (simison) [inline webpack.config.js]: `_.compact` can be now removed

**Comment by simison** [comment]
>Notably in this example

https://github.com/WordPress/gutenberg/blob/1e8d78189a9b05628b171016e102e2c50e267c39/lib/packages-dependencies.php#L195-L204

Also in Gutenberg `master` branch.

**Comment by gziolo** [comment]
> I've built Gutenberg with this plugin [using the instructions above](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946), full results below are very interesting to compare with [the declared script deps in core](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L268-L494).

The list of deps is maintained manually in PHP file so I'm not surprised seeing some differences in there. We would love to have something like that in Gutenberg. If webpack could regenerate `packages-dependencies.json` file every time it is built, it would be amazing. One file less to maintain for us.

**Comment by sirreal** [comment]
Another interesting example from Gutenberg:

[`editor` depends on React](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/js/dist/editor.js#L199) via TextareaAutosize.

---

> We would love to have something like that in Gutenberg.

I think this would provide value to a larger part of the community and would be interested in moving it into Gutenberg.

I've addressed initial feedback and removed the static module mapping. `@wordpress/` deps are now dynamically externed and added to the `.deps.json` files. I've also reworked some of the webpack logic, in particular the `.deps.js` files are based on how the `SourceMapDevtool` plugin generates its assets.

I'd welcome another round of reviews at this time.

File: packages/wordpress-external-dependencies-plugin/README.md
Comment (sirreal) [inline packages/wordpress-external-dependencies-plugin/README.md]:
```suggestion
module.exports = {
```

File: packages/wordpress-external-dependencies-plugin/package.json
Comment (sirreal) [inline packages/wordpress-external-dependencies-plugin/package.json]:
```suggestion
	"homepage": "https://github.com/Automattic/wp-calypso/packages/wordpress-external-dependencies-plugin",
```

File: packages/wordpress-external-dependencies-plugin/package.json
Comment (sirreal) [inline packages/wordpress-external-dependencies-plugin/package.json]:
```suggestion
	"homepage": "https://github.com/Automattic/wp-calypso/tree/master/packages/wordpress-external-dependencies-plugin",
```

**Comment by sirreal** [comment]
Rebased to fix conflicts.

**Review by simison — APPROVED** [review]
Looks good and tested that this works well with SDK builder, too:

```
npm run sdk -- generic --global-wp packages/o2-blocks/src/editor.js ./extern-test.js
```

**Review by simison — CHANGES_REQUESTED** [review]
Actually one thing I just noticed;

in this branch, building with `calypso-build` gives me a new cache folder that isn't gitignored:

```
packages/calypso-build/.cache
```

Just run `npx lerna run prepare --scope="*/o2-blocks"` to replicate.

**Comment by sirreal** [comment]
> in this branch, building with calypso-build gives me a new cache folder that isn't gitignored:

I believe that from a stale branch point. It was fixed on master by #32040.

**Comment by simison** [comment]
Confirmed, all good in `master`. Thanks! :-)

**Pull request merged**
